### PR TITLE
feat(llama-cpp-server): reuse local server for completion and chat

### PR DESCRIPTION
**Pull request by zwpaper**
fix https://github.com/TabbyML/tabby/issues/2654

### CODE REVIEW

File: crates/llama-cpp-server/src/lib.rs [hunk 1]
```diff
@@ -202,6 +202,32 @@ pub async fn create_completion(
     (stream, prompt_info)
 }
 
+pub async fn create_completion_and_chat(
```
Comment (wsxiaoys) [inline crates/llama-cpp-server/src/lib.rs:205]: This seems to still start two llama-server subprocesses, not actually reusing the process?

Comment (zwpaper) [reply crates/llama-cpp-server/src/lib.rs:205]: still working on it, I first changed the outside logic, it's nearly done.
But it's too late last night, so I created this draft, will fix it today.

File: crates/llama-cpp-server/src/lib.rs [hunk 2]
```diff
@@ -89,6 +89,10 @@ impl CompletionServer {
             context_size,
         );
         server.start().await;
+        Self::new_with_server(Arc::new(server)).await
+    }
+
+    async fn new_with_server(server: Arc<LlamaCppSupervisor>) -> Self {
```
Comment (wsxiaoys) [inline crates/llama-cpp-server/src/lib.rs:95]:
```suggestion
    async fn new_with_supervisor(server: Arc<LlamaCppSupervisor>) -> Self {
```

nit: to be consistent with struct naming

File: crates/llama-cpp-server/src/lib.rs [hunk 3]
```diff
@@ -202,6 +210,53 @@ pub async fn create_completion(
...
+        chat_model.parallelism,
+        Some(chat_template),
+        chat_model.enable_fast_attention.unwrap_or_default(),
+        chat_model.context_size,
+    ));
+    server.start().await;
+
+    let chat = ChatCompletionServer::new_with_server(server.clone()).await;
+
+    let completion = if completion_model.model_id == chat_model.model_id {
```
Comment (wsxiaoys) [inline crates/llama-cpp-server/src/lib.rs:244]: also compare other model settings (two LocalModelConfig shall be exactly same other than the `chat_template` field)

File: crates/tabby/src/services/completion.rs [hunk 4]
```diff
@@ -352,26 +354,33 @@ impl CompletionService {
     }
 }
 
-pub async fn create_completion_service(
+pub async fn create_completion_service_and_chat(
```
Comment (wsxiaoys) [inline crates/tabby/src/services/completion.rs:357]:
```suggestion
pub async fn create_completion_and_chat_service(
```

File: crates/tabby/src/services/completion.rs [hunk 5]
```diff
@@ -352,26 +354,33 @@ impl CompletionService {
...
-        logger,
-        prompt_template,
-    )
+    completion: Option<ModelConfig>,
+    chat: Option<ModelConfig>,
+) -> (
+    Option<CompletionService>,
+    Option<Arc<dyn ChatCompletionStream>>,
+) {
+    let (engine, chat, prompt) = model::load_code_generation_and_chat(completion, chat).await;
```
Comment (wsxiaoys) [inline crates/tabby/src/services/completion.rs:367]:
```suggestion
    let (code_generation, chat, prompt) = model::load_code_generation_and_chat(completion, chat).await;
```

File: crates/tabby/src/services/completion.rs [hunk 6]
```diff
@@ -352,26 +354,33 @@ impl CompletionService {
...
+
+    let svc = if let Some(engine) = engine {
```
Comment (wsxiaoys) [inline crates/tabby/src/services/completion.rs:369]:
```suggestion
    let completion = if let Some(engine) = engine {
```

File: crates/llama-cpp-server/src/lib.rs [hunk 7]
```diff
@@ -202,6 +210,53 @@ pub async fn create_completion(
     (stream, prompt_info)
 }
 
+pub async fn create_completion_and_chat(
+    completion_model: &LocalModelConfig,
+    chat_model: &LocalModelConfig,
+) -> (
+    Arc<dyn CompletionStream>,
+    Arc<dyn ChatCompletionStream>,
+    PromptInfo,
```
Comment (wsxiaoys) [inline crates/llama-cpp-server/src/lib.rs:219]:
```suggestion
    Arc<dyn CompletionStream>,
    PromptInfo,
    Arc<dyn ChatCompletionStream>,
```

**Review by wsxiaoys — COMMENTED** [review]
Mostly nits, otherwise LGTM

**Pull request merged**